File: 0242. Valid Anagram.js
// INPUT: "anagram", "nagaram"
// OUTPUT: true

const isAnagramLinearSpace = (str1, str2) => {
  const getChars = str => {
    const chars = {};

    for (let i = 0; i < str.length; i++) {
      if (chars[str[i]]) {
        chars[str[i]]++;
      } else {
        chars[str[i]] = 1;
      }
    }

    return chars;
  };

  const chars1 = getChars(str1),
    chars2 = getChars(str2);

  for (let i of Object.keys(chars1)) {
    if (chars1[i] !== chars2[i]) return false;
  }

  for (let i of Object.keys(chars2)) {
    if (chars1[i] !== chars2[i]) return false;
  }

  return true;
};

const isAnagram = (str1, str2) => {
  const charsCount = new Array(26).fill(0);

  for (let i = 0; i < str1.length; i++) {
    charsCount[str1.charCodeAt(i) - "a".charCodeAt(0)]++;
  }

  for (let i = 0; i < str2.length; i++) {
    charsCount[str2.charCodeAt(i) - "a".charCodeAt(0)]--;
  }

  for (let i = 0; i < 26; i++) {
    if (charsCount[i]) return false;
  }

  return true;
};

console.log(isAnagram("anagram", "nagaram"));

File: python/0242. Valid Anagram.py
class Solution:
    def isAnagram(self, str_1: str, str_2: str) -> bool:
        if len(str_1) != len(str_2):
            return False

        chars = {}

        for idx, _ in enumerate(str_1):
            if str_1[idx] not in chars:
                chars[str_1[idx]] = 0

            chars[str_1[idx]] += 1

            if str_2[idx] not in chars:
                chars[str_2[idx]] = 0

            chars[str_2[idx]] -= 1

        for val in chars.values():
            if val != 0:
                return False

        return True
